### PR TITLE
Refactor so that wrapping functionality can be reused more easily

### PR DESCRIPTION
**Pull request by alex-lee**
I wanted to incorporate black-macchiato into a plugin, so I refactored the `macchiato` function to expose the wrapping functionality. I'm not sure if this is the sort of change you'd want to merge, but in case it looks worthwhile, here it is :slightly_smiling_face:

### CODE REVIEW

File: macchiato.py [hunk 1]
```diff
@@ -2,102 +2,168 @@
...
 
+_single_indent = " " * 4
 
-def macchiato(in_fp, out_fp, args=None):
-    if args is None:
-        args = []
 
-    # Read input.
-    lines = in_fp.readlines()
+class WrapInfo(NamedTuple):
```
Comment (wbolster) [inline macchiato.py:16]: can't say i like namedtuples in general, but i guess using `dataclass` means this code won't work with python3.6 anymore...

File: macchiato.py [hunk 2]
```diff
@@ -2,102 +2,168 @@
...
 import tempfile
 import tokenize
+from typing import IO, Iterable, List, NamedTuple, Optional, Tuple, cast
 
 import black
 
 
 __version__ = "1.2.0"
 
+_single_indent = " " * 4
```
Comment (wbolster) [inline macchiato.py]:
```suggestion
SINGLE_INDENT = " " * 4
```

File: macchiato.py [hunk 3]
```diff
@@ -2,102 +2,168 @@
...
+
+def _fake_before_lines(first_line: str) -> List[str]:
+    """Construct the fake lines that should go before the text."""
+
+    fake_lines = []
+    indent_levels = _indent_levels(first_line)
+
+    # Handle regular indent
+    for i in range(indent_levels):
+        prefix = _single_indent * i
```
Comment (wbolster) [inline macchiato.py]:
```suggestion
        prefix = SINGLE_INDENT * i
```

File: macchiato.py [hunk 4]
```diff
@@ -2,102 +2,168 @@
...
     try:
-        first_token = next(
+        first_token: Optional[tokenize.TokenInfo] = next(
             tokenize.generate_tokens(iter([first_line.lstrip()]).__next__)
         )
     except tokenize.TokenError:
         first_token = None
     if first_token and first_token.type == tokenize.NAME:
         name = first_token.string
+        prefix = _single_indent * indent_levels
```
Comment (wbolster) [inline macchiato.py]:
```suggestion
        prefix = SINGLE_INDENT * indent_levels
```

File: macchiato.py [hunk 5]
```diff
@@ -2,102 +2,168 @@
...
         if name in {"else", "elif"}:
-            lines.insert(n_fake_before, f"{indent * ' '}if True:\n")
-            lines.insert(n_fake_before + 1, f"{indent * ' '}    pass\n")
-            n_fake_before += 2
+            fake_lines.append(f"{prefix}if True:\n")
+            fake_lines.append(f"{prefix}{_single_indent}pass\n")
```
Comment (wbolster) [inline macchiato.py]:
```suggestion
            fake_lines.append(f"{prefix}{SINGLE_INDENT}pass\n")
```

File: macchiato.py [hunk 6]
```diff
@@ -2,102 +2,168 @@
...
-
-    # Detect an unclosed block at the end. Add ‘pass’ at the end of the line if
-    # needed for valid syntax.
-    last_line = lines[-1]
-    n_fake_after = 0
-    if last_line.rstrip().endswith(":"):
-        lines[-1] = last_line.rstrip() + "pass\n"
-        n_fake_after = 1
+            fake_lines.append(f"{prefix}try:\n")
+            fake_lines.append(f"{prefix}{_single_indent}pass\n")
```
Comment (wbolster) [inline macchiato.py]:
```suggestion
            fake_lines.append(f"{prefix}{SINGLE_INDENT}pass\n")
```

File: macchiato.py [hunk 7]
```diff
@@ -2,102 +2,168 @@
...
+        trailing_fake_pass=trailing_fake_pass,
+    )
+    return lines, wrap_info
+
+
+def format_lines(lines: List[str], black_args=None) -> List[str]:
+    if black_args is None:
+        black_args = []
+
+    with tempfile.NamedTemporaryFile(suffix=".py", mode="wt+", delete=False) as fp:
```
Comment (wbolster) [inline macchiato.py]: while at it, maybe write the temp file file in the `os.getcwd()`? see my comments in #4.

**Review by wbolster — CHANGES_REQUESTED** [review]

**Comment by wbolster** [comment]
hey @alex-lee thanks for your interest. happy to merge this; left a few comments.

this whole project started out as a really quick 'n dirty hack, and hence was just the simplest ‘stupid’ script that could work... but then i added support for new corner cases, and other people added more... :see_no_evil:

and while i'm not really comfortable treating this code as a ‘library’ (instead of a command line tool), in practice the changes needed for that are not that big anyway, and this pull request does that in a more than decent way (docstrings, type annotations, various cleanups, and so on). so yeah, :shipit: :+1:

File: macchiato.py (hunk 7 above)
Comment (alex-lee) [reply macchiato.py]: Ok, I've made the change for this and added some tests to see if the config file is detected properly.

**Comment by alex-lee** [comment]
Thanks for taking a look @wbolster and for the comments! Let me know if the working directory change and the related tests look ok.

**Review by wbolster — APPROVED** [review]
this looks great, thanks!

**Comment by wbolster** [comment]
fyi i pushed 1.3.0 with these changes to pypi: https://pypi.org/project/black-macchiato/1.3.0/

**Comment by wbolster** [comment]
heh and looking at that page it reminded me of https://github.com/wbolster/black-macchiato/issues/3  :see_no_evil:

**Pull request merged**